Create a coupon block in the email editor [MAILPOET-4678]

samnajian opened the pull request:
## Description

This PR adds simple the Coupon block and takes care of the following:
1) Adding the block to the email
2) Adding default config
3) Generating a coupon for each added block and then persisting the coupon id along with the block settings
4) Generating a coupon for `automatic` emails but not persisting the coupon id with the block so the the next rendering/send out of the email gets a new coupon.



## QA notes:
- The feature is behind a feature flag
- Please add the coupon to and email and check if the coupon is generated when it's send out and not in preview
- In preview we show only a placeholder of the coupon, unless the newsletter already has a coupon generated and assigned to it
- The validation as of now only happens on frontend as this is the simple implementation of the feature and more validation are to be added (especially on backend side) in other tickets
- The only email time that will get a dynamic coupon on sending (dynamic means the coupon is not saved to the coupon block setting, but only generated once the email is sent out), are the `automatic` email types, one example of those are `abandoned cart` emails, the rest of the email types, get a coupon code generated on the first send out, and for the following send outs the same coupon code will be used
- The coupon code that's generated should have the email subject and ID as description of the coupon, and is viewable under WooCommerce -> Coupons menu.
- 
## Linked tickets

[MAILPOET-4678]




[MAILPOET-4678]: https://mailpoet.atlassian.net/browse/MAILPOET-4678?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

samnajian committed:
Bootstrap Coupon block and make it function

[MAILPOET-4678]

samnajian committed:
Add Coupon block setting pane

[MAILPOET-4678]

samnajian committed:
Add coupon block config

[MAILPOET-4678]

samnajian committed:
Adjust coupon block styling in editor

[MAILPOET-4678]

samnajian committed:
Render Coupon block in email body

[MAILPOET-4678]

samnajian committed:
Remove coupon code input from the settings sidebar

[MAILPOET-4678]

samnajian committed:
Add information overlay to Coupon block

[MAILPOET-4678]

samnajian committed:
Add CouponPreprocessor to create coupons before rendering

[MAILPOET-4678]

samnajian committed:
Pass Coupon code placeholder from PHP and unify

[MAILPOET-4678]

samnajian committed:
Add WC helper function to retrieve coupon code from id

[MAILPOET-4678]

samnajian committed:
Generate coupon code for each coupon block missing the code

[MAILPOET-4678]

samnajian committed:
Set missing parameters when creating coupon

[MAILPOET-4678]

samnajian committed:
Add functionality to update coupon when block settings change

[MAILPOET-4678]

samnajian committed:
Make sure coupon code placeholder doesn't generate coupon

[MAILPOET-4678]

samnajian committed:
Persist coupon block expiration in WC_Coupon

[MAILPOET-4678]

samnajian committed:
Allow coupon generation per render for automatic emails

[MAILPOET-4678]

costasovo commented on mailpoet/assets/js/src/newsletter_editor/blocks/coupon.ts:
```diff
@@ -0,0 +1,213 @@
+  initialize(...args) {
+    base.BlockView.prototype.initialize.apply(this, args);
+
+    // Listen for attempts to change all dividers in one go
```
Shouldn't `all dividers` be `all coupons` here?

costasovo commented on mailpoet/assets/js/src/global.d.ts:
```diff
@@ -183,4 +183,5 @@ interface Window {
     subscribers_count: number;
   };
   templates?: Record<string, string>;
+  is_wc_active?: boolean;
```
We already have that on window see https://github.com/mailpoet/mailpoet/blob/a6bb3809ca363e2902c2eb42602845f266c0358f/mailpoet/assets/js/src/global.d.ts#L92
This one is also available on MailPoer modude under `MailPoet.isWoocommerceActive`.

Could you please use the exisitng value. Also when adding a new global value on window please prefix it with `mailpoet_` to prevent potential conflicts. 

Note: I noticed that there are also `window.templates` without the prefix. I see we have it like that historically. Would be nicer to have `mailpoet_templates` but I would say the change is out of the scope of this change.

samnajian replied:
nice catch, that's a left over, will get removed.

samnajian replied:
I was also not comfortable with keeping the name `template` but wanted to refrain from adding even more changes.

costasovo replied:
Let's keep the template as is. It has been like that for ages.

costasovo commented on mailpoet/lib/Newsletter/Renderer/Blocks/Coupon.php:
```diff
@@ -0,0 +1,59 @@
+<?php // phpcs:ignore SlevomatCodingStandard.TypeHints.DeclareStrictTypes.DeclareStrictTypesMissing
```
It is a new file so please add ` declare(strict_types = 1);` instead of the ignore comment for the code sniffer

costasovo commented on mailpoet/lib/WooCommerce/CouponPreProcessor.php:
```diff
@@ -0,0 +1,60 @@
+      substr($chars, rand(0, $length - 5), 4)
+    );
+  }
+}
```
Just a suggestion. We already have a util method for generating a random code. See https://github.com/mailpoet/mailpoet/blob/926620e8f82ca691f1c11f6e973a2f86a6117522/mailpoet/lib/Util/Security.php#L36 so maybe you could use this one instead of implementing own.

samnajian replied:
I have no strong opinion about this, fine with using `generateRandomString` if you think that's better.

costasovo replied:
Using `generateRandomString` makes sense to me so that we don't re-create similar functionality we already have.

costasovo commented on mailpoet/lib/WooCommerce/CouponPreProcessor.php:
```diff
@@ -0,0 +1,60 @@
+    $code = $couponCode ?? $this->generateRandomCode();
+    $coupon->set_code($code);
+
+    return $coupon->save();
```
What happens if we accidentally generate an existing code? Will `coupon->save` throw an exception or return an error? I just want to ensure we don't end up with an error in preprocessing that would block sending.

From what I experienced when testing, when we set coupon type percent and coupon value more than 100, there will be an error.

samnajian replied:
That's correct, it'll throw, I was under the impression to leave this to the validation ticket, but your `it won't get released under a feature flag` is valid and I'll manage this.

costasovo commented on mailpoet/lib/WooCommerce/CouponPreProcessor.php:
```diff
@@ -65,6 +66,8 @@ private function addOrUpdateCoupon(array $couponBlock, NewsletterEntity $newslet
     }
     $coupon->set_discount_type($couponBlock['discountType']);
     $coupon->set_amount($couponBlock['amount']);
+    $expiration = (new DateTime())->getCurrentDateTime()->modify("+{$couponBlock['expiryDay']} day")->getTimestamp();
+    $coupon->set_date_expires($expiration);
```
Just want to double-check check I understand correctly. This updates the expiration each time when the plugin pre-process a newsletter. So for the welcome email, it prolongs expiration every time when the newsletter is sent. Correct? I don't think it is a big issue. Just want to make sure we prolong the expiration so that the new subscribers have a valid coupon. Maybe an integration test would be nice for that.

samnajian replied:
yes, so basically the actual expiration is the one of the last send out, adding a test makes sense.

costasovo replied:
Adding test is a good idea. I think it would be nice to cover whole class with an integration test. It seems that the class handles all the logic about coupon creation. I assume we will be adding more in the follow-up tasks.

costasovo commented on mailpoet/lib/WooCommerce/CouponPreProcessor.php:
```diff
@@ -91,4 +91,11 @@ private function generateRandomCode(): string {
+   * For some renders/send outs the coupon id shouldn't be persisted along the coupon block
+   */
+  private function shouldPersist(NewsletterEntity $newsletter): bool {
+    return $newsletter->getType() !== NewsletterEntity::TYPE_AUTOMATIC;
```
Aha, so that means that for automatic emails, we will generate a coupon for every sending? Correct?
I think that TYPE_AUTOMATION emails and maybe also TYPE_WELCOME potentially also  TYPE_RE_ENGAGEMENT and TYPE_CONFIRMATION_EMAIL_CUSTOMIZER have the same behavior that they can be sent after some event.

samnajian replied:
The implementation of this method is kept simple, mostly like a placeholder, since we have a ticket to dynamically create coupons for the newsletters you mentioned ( or even more?)

costasovo replied:
Aha. If it is a placeholder, maybe we could always return true and add a comment that in the future, it will return true for dynamic coupons. Or maybe just add a comment. When I saw the code it was suspicious. Why only automatic emails?

costasovo reviewed (changes requested):
Thanks for implementing this feature! Good job! This is a really big feature.

I've added some comments. 

I run into an issue when testing:

It is possible to add random values for amount. For example, when wanted to use the percent coupon type and passed value higher than 100 the sending failed with and error `Invalid discount amount`. I'm not sure but I think that the error was coming grom Woo code for saving a coupon. I was testing with a welcome email and the email ended up stuck in sending state, which is an issue because this would block the background job. We need to make sure we can recover when there is an error when saving the coupon and also, it would be nice to restrict the amount to numbers only and don't allow more than 100 for percent coupons. I noticed in the ticket comments that some validation will be done in follow-up ticket. But since this is not hidden behind a feature flag we should at least make sure that invalid values will not block sending.

<img width="336" alt="Screenshot 2023-01-13 at 11 18 44" src="https://user-images.githubusercontent.com/1082140/212321380-3a24c788-4320-42e9-a9bd-e890131826a8.png">

I would also suggest adding some tests for the new functionality:
1) The CouponPreProcessor contains a lot of logic that I think really needs an integration test
2) We have [unit tests for block renderers](https://github.com/mailpoet/mailpoet/tree/82aeb898540e726ec6a5d177d70d9031202d1e06/mailpoet/tests/unit/Newsletter/Renderer/Blocks) so I would consider adding test for this block as well
3) We have ["component level" test for newsletter editor blocks](https://github.com/mailpoet/mailpoet/tree/d116ff360b8d73f0ae09429ae0c365ed30a54def/mailpoet/tests/javascript_newsletter_editor/newsletter_editor/blocks). I would suggest adding tests for the new block.
4) I would add a very basic acceptance test (checking block can be added to the email content, setting panel can be opened and some value can be set and coupon saved)

samnajian committed:
Fix mistake in comment

[MAILPOET-4678]

samnajian committed:
Add missing strict_types to Coupon.php

[MAILPOET-4678]

samnajian committed:
Move Coupon instantiation to MailPoet\WooCommerce\Helper

[MAILPOET-4678]

samnajian committed:
Properly check for existance of code in coupon block setting

[MAILPOET-4678]

samnajian committed:
Add unit tests for CouponPreProcessor

[MAILPOET-4678]

samnajian committed:
Update coupon setting view styling

Adds percentage sign when the discount type is of percentage
type, adjusts the width and view of coupon amount field

[MAILPOET-4678]

samnajian committed:
Adjust styling and add validation to coupon amount

[MAILPOET-4678]

samnajian committed:
Add validation to expiry days field of coupon block

[MAILPOET-4678]

samnajian committed:
Add more unit tests for CouponPreProcessor

[MAILPOET-4678]

samnajian committed:
Add Coupon block feature flag

[MAILPOET-4678]

samnajian committed:
Augment comment for CouponPreProcessor::shouldPersist

[MAILPOET-4678]

samnajian committed:
Add newsletter id to the coupon description

[MAILPOET-4678]

costasovo commented on mailpoet/tests/unit/WooCommerce/CouponPreProcessorTest.php:
```diff
@@ -24,6 +25,10 @@ public function __construct(
   ) {
     parent::__construct($name, $data, $dataName);
 
+    WordPress::interceptFunction('wp_timezone', function() {
+      new \DateTimeZone('UTC');
+    });
```
Hmm, I was not aware we had this! TIL 👍

costasovo reviewed (dismissed):
Thanks for adding the validation and the CouponProcessor test.

Aschepikov commented:
Hi, @samnajian I found some issues, some of them can be connected with block settings or will be fixed in the next pr, take a look at them pls.

1. After coupon adding there is a console error
![image](https://user-images.githubusercontent.com/113185437/213670800-19f6e567-477d-4b78-9cd3-08a12eb23b9d.png)

2. The percent symbol is moving with scrolling
![image](https://user-images.githubusercontent.com/113185437/213671134-5d488fc6-2f21-4858-a661-7fbb99db06a7.png)
![image](https://user-images.githubusercontent.com/113185437/213671149-fe3017cd-629a-4ad8-a1fd-e9c4ea47fa8e.png)

3. I can see just a code of the colors but usually we can see color box like in the button settings and we have an issue with the border position
![image](https://user-images.githubusercontent.com/113185437/213671314-bca83830-71f3-463d-92f1-d2a4a3184d57.png)
In the button settings it is like
![image](https://user-images.githubusercontent.com/113185437/213671282-b69dc318-6c93-4296-a760-8a702efc52c6.png)

4. Issue with the positioning of the text when there is a long name of font
![image](https://user-images.githubusercontent.com/113185437/213671500-fcf838fd-d1c8-4a28-9d26-57a987371f6f.png)
![image](https://user-images.githubusercontent.com/113185437/213671553-6bb74b0e-0ee9-4cb4-970c-5e66680e2c14.png)
It can be connected with the whole block I can make a ticket
![image](https://user-images.githubusercontent.com/113185437/213671424-7161c3ae-59de-420a-852e-4f0169a74aaf.png)

5. When we are changing from the fixed cart discount to the fixed product discount and back, the coupon amount field is green, but with percentage discount is white, why
![image](https://user-images.githubusercontent.com/113185437/213671771-2e760346-b43a-4ca6-9146-c8e1a0730573.png)
![image](https://user-images.githubusercontent.com/113185437/213671791-cc4f9e41-30f5-488c-aae9-afc147a3e33e.png)
![image](https://user-images.githubusercontent.com/113185437/213671805-15da3ce3-7285-4945-8859-7a9333d7c73f.png)
![image](https://user-images.githubusercontent.com/113185437/213671826-8581ee2c-aa1e-4e10-ae1c-aa4911bb0982.png)
And every time when we are changing the coupon amount within validation it is green but why it is green after changing from the fixed cart discount to the fixed product discount and back

6. Coupon amount is not changing when we are changing the discount type so we can set an amount which is not validated and we don't see validation error 
![image](https://user-images.githubusercontent.com/113185437/213671896-7277c2e7-5cdb-4e86-96b5-97a7de85866f.png)
![image](https://user-images.githubusercontent.com/113185437/213671921-10b58452-fd14-46d9-81f7-75e862a4946f.png)
With wrong validation after that, the sending of the email doesnt working 
![image](https://user-images.githubusercontent.com/113185437/213672132-bcf8b2b2-180e-4435-b2d1-29cd51ab8180.png)

7. There are no font changes on the preview or in the email
![image](https://user-images.githubusercontent.com/113185437/213672281-3bc75d45-15f7-44a8-b09d-6f6216ea5eb5.png)
![image](https://user-images.githubusercontent.com/113185437/213672295-15ffb28b-8f11-4d0e-bf69-fac56d2ff679.png)

8. I can't send the second newsletter with coupons, just to make copies of the first, and even if I will change some data in the coupon settings it will not change them, but if I will make a copy and delete all of the coupons and make them again is sending much longer and has the same coupon code but new data
![image](https://user-images.githubusercontent.com/113185437/213672371-be58d9b8-b32d-4436-a5a8-4f0ec13da7da.png)
![image](https://user-images.githubusercontent.com/113185437/213672387-40e0e390-2586-4bdb-af06-232c69a97ab2.png)
![image](https://user-images.githubusercontent.com/113185437/213672413-f61a944a-d076-461d-86b0-e913e8493472.png)
![image](https://user-images.githubusercontent.com/113185437/213672430-05514b5a-fd9f-4759-920a-38d558d3f6e9.png)
![image](https://user-images.githubusercontent.com/113185437/213672454-7ca65f22-4916-4f79-9356-47060810dbeb.png)

samnajian committed:
Fix bug in validating on render

[MAILPOET-4678]

samnajian committed:
Fix bug causing the percent sign to move on scroll

[MAILPOET-4678]

samnajian commented:
@Aschepikov Thanks for the quick feedback, much appreciated, some of the issues you reported were because of the js error on render, that’s fixed and pushed now, so please give it another check and those rendering issues are gone, apart from that, please note that the none-coupon settings are 100% the same as the ones for the button block, so the issues are the same ( things like longer font name causing the font-size selector to get cut ) and we’ll be tackling the design in a separate ticket anyways, so you can ignore these

samnajian committed:
Adjust percent sign positioning

[MAILPOET-4678]

samnajian committed:
Re-adjust percent sign positioning

[MAILPOET-4678]

samnajian committed:
Update translators hint for the coupon description

[MAILPOET-4678]

samnajian committed:
Cleanup coupon sensitive data from blockDefaults

[MAILPOET-4678]

samnajian committed:
Fix bug when duplicated newsletter uses the source's coupon

On duplication, the sensitive data of the coupon were
inherited from the source

[MAILPOET-4678]

samnajian committed:
Replace coupon code with code placeholder when duplicationg

[MAILPOET-4678]

samnajian committed:
Fix failing unit test

[MAILPOET-4678]

costasovo commented on mailpoet/assets/js/src/newsletter_editor/components/save.js:
```diff
@@ -53,6 +53,8 @@ Module.save = function () {
       delete json.body.blockDefaults.woocommerceHeading;
       delete json.body.blockDefaults.woocommerceContent;
+      delete json.body.blockDefaults?.coupon.couponId;
+      delete json.body.blockDefaults?.coupon.code;
```
I think here we should also check the coupon is set `delete json.body.blockDefaults?.coupon?.code`;

samnajian committed:
Change NewsletterCoupon::cleanupSensitiveData signature

This change allows passing body array and retuning body array
so that NewsletterEntity instance doesn't need to get passed
around

[MAILPOET-4678]

samnajian committed:
Cleanup coupon data from template on save

[MAILPOET-4678]

samnajian committed:
Check blockDefaults.coupon when trying to delete its props

[MAILPOET-4678]

costasovo reviewed (dismissed):
LGTM! 👍

samnajian committed:
Fix issue in test where string was set to newsletter body

[MAILPOET-4678]

Aschepikov reviewed (approved):
QA verified
